A36-xds-for-servers.md

```diff
@@ -0,0 +1,401 @@
+fail racing RPCs but in a way that the client may transparently retry. The
+XdsServer is free to use a different "not serving" strategy post-startup than
+for the initial startup.
+
+The XdsServer does not have to wait until server credentials (e.g., TLS certs)
+are available before accepting connections; since XdsServerCredentials might not
+be used, the server is free to lazily load credentials. However, the XdsServer
+should keep the credentials cache fresh and up-to-date after that initial
+lazy-loading, as it is clear at that point that XdsServerCredentials are being
+used.
+
+The XdsServer API will allow applications to register a "serving state" callback
+to be invoked when the server begins serving and when the server encounters
+errors that force it to be "not serving". If "not serving", the callback must be
+provided error information, for debugging use by developers. The error
+information should generally be language-idiomatic, but determining the cause of
+the error does not need to be machine-friendly. If the application does not
+register the callback, XdsServer should log any errors and each serving
+resumption after an error, all at a default-visible log level.
+
+XdsServer's start must not fail due to transient xDS issues, like missing xDS
+configuration from the xDS server. If XdsServer's start blocks waiting for xDS
+configuration an application can use the serving state callback to be notified
+of issues preventing startup progress.
+
+### xDS Protocol
+
+The `GRPC_XDS_BOOTSTRAP` file will be enhanced to have a new field:
+```
+{
+  // A template for the name of the Listener resource to subscribe to for a gRPC
+  // server. If the token `%s` is present in the string, it will be replaced
+  // with the server's listening "IP:port" (e.g., "0.0.0.0:8080", "[::]:8080").
+  "server_listener_resource_name_template": "example/resource/%s",
+  // ...
+}
+```
+
+XdsServer will use the normal `XdsClient` to communicate with the xDS server.
+There is no default value for `server_listener_resource_name_template` so if it
+is not present in the bootstrap then server creation or start will fail or the
+XdsServer will become "not serving". XdsServer will perform the `%s` replacement
+in the template (if the token is present) to produce a listener name. The
+XdsServer will start an XdsClient watch on the listener name for a
+`envoy.config.listener.v3.Listener` resource. No special character handling of
+the template or its replacement is performed. For example, with an address of
+`[::]:80` and a template of `grpc/server?xds.resource.listening_address=%s`, the
+resource name would be `grpc/server?xds.resource.listening_address=[::]:80`.
+
+To be useful, the xDS-returned Listener must have an
+[`address`][Listener.address] that matches the listening address provided. The
+Listener's `address` would be a TCP `SocketAddress` with matching `address` and
+`port_value`. The XdsServer must be "not serving" if the address does not match.
```

Comment:
This might be more of an implementation related question. In Go, the layer in the `xdsClient` which handles the processing and parsing of the raw message received from the control plane (and makes the decision whether to ACK or NACK), is pretty low and does not have an idea about things like the resource name template configured in the bootstrap file. Currently this layer extracts the listening_address encoded as `ip:port` in the resource name and compares it against the socket address in the `address`. If they don't match, it rejects the resource. And once the resource is rejected here (and NACK'ed), upper layers (the ones which register watches on these resources) like the `xdsServer` will not even be notified about this. This means that an `xdsServer` will continue to remain blocked  waiting for the first resource or will continue using an earlier received good listener resource.

The question here is whether we can NACK the update when the listening address doesn't match and stop there or do we *have* to move the `xdsServer` to "not serving" state?
- This is not much of a concern if this issue happens as part of the first update since the server is anyways blocked on receiving the first good update and can continue to do so.
- But when the server is already using a good configuration and gets an update where the address doesn't match, it does raise an important question.

From the section on serving vs not serving, it does seem to be clear that we need to move to not-serving 
```
However, XdsServer must accept configuration changes provided by
the xDS server, including resource deletions. If that causes the XdsServer to
lack essential configuration (e.g., the Listener was deleted) the server would
need to enter "not serving" mode ...
```

How is Java/C++ handling this? @sanjaypujare @yashykt 

On an orthogonal note, the changes that we plan to make for CSDS which will ensure that all resources are parsed (even if some error out), and that upper layers are notified might help simplify things here.

Reply:
> If they don't match, it rejects the resource

No, it does not reject (NACK) the resource. The XdsClient validation is context-independent, so it won't be able to validate the listening_address is for the address the XdsServer needs. That was something we lost when we swapped away from the "custom xds.resource.listening_address protocol." If there is an address mismatch, XdsClient will ACK the resource but XdsServer will go "not serving." Not a great situation to be in, but it isn't the only time that sort of thing can happen and there's not really any alternatives.

I agree that _if we NACKed_, then XdsServer should just remain in whatever state it was previously (which would be easy). But we don't NACK.